#53285 switch to use MemberInfo.MetadataToken

FatTigerWang opened the pull request:
Switch to use MemberInfo.MetadataToken

With #53283, we won't build a netstandard1.x configuration of this library anymore and the code path can be optimized.

Fix #53285

FatTigerWang committed:
#53285 switch to use MemberInfo.MetadataToken

FatTigerWang committed:
fix code style

fix code style

FatTigerWang committed:
remove whitespace

remove whitespace to fix SA1028

krwq commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -831,21 +831,10 @@ public sealed override bool Equals(MethodInfo? left, MethodInfo? right)
-                        return false;
-
-                    if (left.Name != right.Name)
+                    if (left.MetadataToken != right.MetadataToken)
```
Is it possible that left is TypeRef and right is TypeDef? that could potentially produce false where it shouldn't.

Also if the token is equal should it just return true? Could arguments differ for same token?

FatTigerWang replied:
As far as I know, MetadataToken is unique in the same Module, so when MetadataToken is not equal, it can return false. In addition, according to [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.memberinfo.metadatatoken?view=net-5.0&viewFallbackFrom=netstandard-1.0#System_Reflection_MemberInfo_MetadataToken), I am not sure whether it is necessary to judge the unexpected situation. Do you have any suggestions? I will change the code later.

FatTigerWang committed:
only use MetadataToken and Module for judgment

delete redundant judgments, only use MetadataToken and Module for judgment

jkotas commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -823,54 +823,12 @@ private sealed class MethodInfoEqualityComparer : EqualityComparer<MethodInfo>
-                    }
-
-                    return true;
+                    return left.Module.Name == right.Module.Name && left.MetadataToken == right.MetadataToken;
```
The modules can have same name, but different identity. This condition would not work correctly in that case.

Should the whole MethodInfoEqualityComparer type be deleted and the dictionary switched to use the default comparer? It is not clean why this implementation is trying to achieve that the default EqualityComparer would not.

FatTigerWang replied:
The default comparator will compare the reference of the object. It cannot be judged correctly when used here, so need to use `MethodInfoEqualityComparer`

jkotas replied:
The default EqualityComparer will call Equals method to compare the types: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs#L135

MethodInfo.Equals should do exactly what the existing code is trying to achieve.

FatTigerWang replied:
I did not find information about `MethodInfo.Equals` override from `Object.Equals`

FatTigerWang commented:
Any feedback?

jkotas commented:
It would be useful to understand why a simple `MethodInfo.Equals` does not work. If it does not work, add comment explaining it. Hopefully, it will also help us figure out how the code can be simplified.

FatTigerWang commented:
> It would be useful to understand why a simple `MethodInfo.Equals` does not work. If it does not work, add comment explaining it. Hopefully, it will also help us figure out how the code can be simplified.

You are right, `MethodInfo.Equals` can be judged correctly, so the default EqualityComparer should be used.

FatTigerWang commented:
> It would be useful to understand why a simple `MethodInfo.Equals` does not work. If it does not work, add comment explaining it. Hopefully, it will also help us figure out how the code can be simplified.

You are right, `MethodInfo.Equals` can be judged correctly, so the default EqualityComparer should be used.

FatTigerWang committed:
use Methodinfo.Equals to judge equality

use Methodinfo.Equals to judge equality

use Methodinfo.Equals to judge equality and remove special comparator.

Fix #453285

steveharter commented:
> You are right, MethodInfo.Equals can be judged correctly, so the default EqualityComparer should be used.

Are there automated tests that verify the original issue and\or that the Equals overload works as expected?

FatTigerWang commented:
> > You are right, MethodInfo.Equals can be judged correctly, so the default EqualityComparer should be used.
> 
> 
> 
> Are there automated tests that verify the original issue and\or that the Equals overload works as expected?

I will check automated tests and add new tests if necessary.

FatTigerWang commented:
I have reviewed the existing automated test cases, and I don’t think there is a need to add new test cases for this change.

FatTigerWang commented:
I read the code again, `MethodInfo.Equals` was supported in `.NET Standard 1.0`, here is just to judge the equality of the methods of a single interface, why was the special `EqualityComparer` implemented at that time?

jkotas commented:
I am not able to find that in history.

jkotas commented:
The special equality comparer was added in https://github.com/dotnet/corefx/pull/9546

https://github.com/dotnet/corefx/pull/9546#discussion_r67718913 has more discussion about. I do not think that the arguments in that discussion are valid. In particular, I do not agree that "The default Equality test for MethodInfo is reference equality, not "semantic equality" is true.

jkotas commented:
@FatTigerWang  Thank you!
